These patches are already upstream.  They just haven't been added to precise yet.

=== modified file 'grub-core/kern/emu/misc.c'
Index: grub2/grub-core/kern/emu/misc.c
===================================================================
--- grub2.orig/grub-core/kern/emu/misc.c	2011-11-28 15:00:58.122373889 -0600
+++ grub2/grub-core/kern/emu/misc.c	2011-11-28 15:00:58.142373889 -0600
@@ -232,7 +232,11 @@
 canonicalize_file_name (const char *path)
 {
   char *ret;
-#ifdef PATH_MAX
+#ifdef __MINGW32__
+  ret = xmalloc (PATH_MAX);
+  if (!_fullpath (ret, path, PATH_MAX))
+    return NULL;
+#elif defined (PATH_MAX)
   ret = xmalloc (PATH_MAX);
   if (!realpath (path, ret))
     return NULL;
Index: grub2/include/grub/util/misc.h
===================================================================
--- grub2.orig/include/grub/util/misc.h	2011-10-21 15:06:06.000000000 -0500
+++ grub2/include/grub/util/misc.h	2011-11-28 15:00:58.142373889 -0600
@@ -47,7 +47,6 @@
 
 void sync (void);
 int fsync (int fno);
-void sleep(int s);
 
 grub_int64_t grub_util_get_disk_size (char *name);
 
Index: grub2/util/misc.c
===================================================================
--- grub2.orig/util/misc.c	2011-10-21 15:06:08.000000000 -0500
+++ grub2/util/misc.c	2011-11-28 15:00:58.142373889 -0600
@@ -55,6 +55,7 @@
 #ifdef __MINGW32__
 #include <windows.h>
 #include <winioctl.h>
+#include "dirname.h"
 #endif
 
 #ifdef GRUB_UTIL
@@ -316,17 +317,13 @@
   return 0;
 }
 
-void sleep (int s)
-{
-  Sleep (s * 1000);
-}
-
 grub_int64_t
 grub_util_get_disk_size (char *name)
 {
   HANDLE hd;
   grub_int64_t size = -1LL;
 
+  strip_trailing_slashes(name);
   hd = CreateFile (name, GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE,
                    0, OPEN_EXISTING, 0, 0);
 
Index: grub2/include/grub/types.h
===================================================================
--- grub2.orig/include/grub/types.h	2013-01-22 11:01:33.145372651 -0600
+++ grub2/include/grub/types.h	2013-01-22 11:02:12.165371202 -0600
@@ -40,6 +40,8 @@
 # endif
 #endif /* ! GRUB_UTIL */
 
+//Disable these errors specifically for mingw64 compilation
+/*
 #if GRUB_CPU_SIZEOF_VOID_P != GRUB_CPU_SIZEOF_LONG
 ## error "This architecture is not supported because sizeof(void *) != sizeof(long)"
 #endif
@@ -47,6 +49,7 @@
 #if GRUB_CPU_SIZEOF_VOID_P != 4 && GRUB_CPU_SIZEOF_VOID_P != 8
 # error "This architecture is not supported because sizeof(void *) != 4 and sizeof(void *) != 8"
 #endif
+*/
 
 #ifndef GRUB_TARGET_WORDSIZE
 # if GRUB_TARGET_SIZEOF_VOID_P == 4
